packages/tpetra/core/ext/TpetraExt_MatrixMatrix_def.hpp

```diff
@@ -521,60 +621,145 @@ add (const Scalar& alpha,
+#endif
+    return Teuchos::rcp_static_cast<crs_matrix_type>(Bprime->add(alpha, *Aprime, beta, CDomainMap, CRangeMap, params));
+  }
+  else if(!matchingColMaps)
+  {
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+    MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("mismatched col map full kernel"))));
+#endif
+    //use kernel that converts col indices in both A and B to common domain map before adding
+    auto Acolmap = Aprime->getColMap()->getMyGlobalIndices();
+    auto Bcolmap = Bprime->getColMap()->getMyGlobalIndices();
+    RCP<typename AddKern::GlobalColIndsArray> globalColinds;
+    AddKern::GlobalCols(
+      Aprime->getLocalMatrix(), alpha, Bprime->getLocalMatrix(), beta, Acolmap, Bcolmap,
+      CRangeMap->getMinGlobalIndex(), Aprime->getGlobalNumCols(), vals, rowptrs, globalColinds);
+    colinds = rcp(new ColIndsArray("C colinds", globalColinds->dimension_0()));
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+    MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("building optimized column map"))));
+#endif
+    CrowMap = Bprime->getRowMap();
+    CcolMap = AddKern::getOptimizedColMap(Aprime->getGlobalNumCols(), *globalColinds, *colinds, comm);
+  }
+  else
+  {
+    auto localA = Aprime->getLocalMatrix();
+    auto localB = Bprime->getLocalMatrix();
+    auto Avals = localA.values;
+    auto Bvals = localB.values;
+    auto Arowptrs = (const RowPtrsArray*) &localA.graph.row_map;
+    auto Browptrs = (const RowPtrsArray*) &localB.graph.row_map;
+    auto Acolinds = localA.graph.entries;
+    auto Bcolinds = localB.graph.entries;
+    if(sorted)
+    {
+      //use sorted kernel
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("sorted entries full kernel"))));
+#endif
+      AddKern::Sorted(Avals, *Arowptrs, Acolinds, alpha, Bvals, *Browptrs, Bcolinds, beta, vals, rowptrs, colinds);
+    }
+    else
+    {
+      //use unsorted kernel
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("mm add unsorted entries full kernel"))));
+#endif
+      AddKern::Unsorted(Avals, *Arowptrs, Acolinds, alpha, Bvals, *Browptrs, Bcolinds, beta, Aprime->getGlobalNumCols(), vals, rowptrs, colinds);
+    }
+    CrowMap = Bprime->getRowMap();
+    CcolMap = Bprime->getColMap();
+  }
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("Tpetar::Crs constructor"))));
+#endif
+  C = rcp(new crs_matrix_type(CrowMap, CcolMap, *rowptrs, *colinds, *vals, params));
+#ifdef HAVE_TPETRA_MMM_TIMINGS
+      MM = rcp(new TimeMonitor(*TimeMonitor::getNewTimer(prefix_mmm + std::string("Tpetra::Crs expertStaticFillComplete"))));
+#endif
+  C->expertStaticFillComplete(CDomainMap, CRangeMap, rcp(new import_type(CDomainMap, CcolMap)), rcp(new export_type(CrowMap, CRangeMap)), params);
```

Comment:
This would have been a good opportunity to use the fused Import constructor, no?